### ISSUE TITLE
1D convolution not supported?

### ISSUE THREAD

**Issue by andreistoian** [CEA-LIST/N2D2#74]
Hello! 

I defined and trained a sound classification network in pyTorch and it includes multiple conv1D layers. It's a very simple topology with several conv1D followed by 2 FC layers (and I do not export the softmax). I exported the model to ONNX and I created an INI file to load it into N2D2.

When I import it in N2D2 I get errors:
- when DefaultModel=Frame_CUDA: CUDNN failure: CUDNN_STATUS_BAD_PARAM (3) in ../src/Cell/ConvCell_Frame_CUDA.cpp:503 on device #0
- when DefaultModel=Frame: Error: ConvCell_Frame: only 2D convolution is supported

Does N2D2 support 1D convolution (kernel shape with number of dimensions = 3) or do I need to make pytorch use 2D convolution with a 1xN filter size?

Regards,

**Comment by olivierbichler-cea** [comment]
Hello,
If the convolution is explicitly 1D in the ONNX model, this is indeed not handled... but the fix should be trivial!

Before I commit it, I would like that you test it first. In the ConvCell_Frame_CUDA constructor, could you add at the very beginning the following:

```
    if (kernelDims.size() == 1) {
        mKernelDims.resize(2, 1);
        mSubSampleDims.resize(2, 1);
        mStrideDims.resize(2, 1);
        mPaddingDims.resize(2, 0);
        mDilationDims.resize(2, 1);
    }
```

Cheers,
Olivier

**Comment by olivierbichler-cea** [comment]
Sorry that's not going to work because of const members.

Could you try to change the ConvCell constructor instead?

```
N2D2::ConvCell::ConvCell(const DeepNet& deepNet, 
                         const std::string& name,
                         const std::vector<unsigned int>& kernelDims,
                         unsigned int nbOutputs,
                         const std::vector<unsigned int>& subSampleDims,
                         const std::vector<unsigned int>& strideDims,
                         const std::vector<int>& paddingDims,
                         const std::vector<unsigned int>& dilationDims)
    : Cell(deepNet, name, nbOutputs),
      mNoBias(this, "NoBias", false),
      mBackPropagate(this, "BackPropagate", true),
      mWeightsExportFormat(this, "WeightsExportFormat", OCHW),
      mWeightsExportFlip(this, "WeightsExportFlip", false),
      mOutputsRemap(this, "OutputsRemap", ""),
      mKernelDims((kernelDims.size() == 1)
        ? std::vector<unsigned int>({kernelDims[0], 1}) : kernelDims),
      mSubSampleDims((subSampleDims.size() == 1)
        ? std::vector<unsigned int>({subSampleDims[0], 1}) : subSampleDims),
      mStrideDims((strideDims.size() == 1)
        ? std::vector<unsigned int>({strideDims[0], 1}) : strideDims),
      mPaddingDims((paddingDims.size() == 1)
        ? std::vector<int>({paddingDims[0], 0}) : paddingDims),
      mDilationDims((dilationDims.size() == 1)
        ? std::vector<unsigned int>({dilationDims[0], 1}) : dilationDims),
      mExtPaddingDims(2 * kernelDims.size(), 0)
{
    // ctor
}
```

**Comment by andreistoian** [comment]
Thanks for your answer. I updated the code as you suggested and also removed the dimension check : 
```
    if (kernelDims.size() != 2) {
        throw std::domain_error("ConvCell_Frame: only 2D convolution is"
                                " supported");
    }
```

But then I run into a problem with a mismatch with the input dimension: 

```
void N2D2::ConvCell::setOutputsDims()
{
    if (mKernelDims.size() != mInputsDims.size() - 1) {
        std::stringstream msgStr;
        msgStr << "ConvCell::setOutputsDims(): the number of dimensions of the"
```

I need to set the input dimensions in the ini file as following in order to pass the check of the size of StimuliProvider with the ONNX model :

```
[sp]
Size=8000 1
```

with "1 8000 1" I get "Unexpected size for ONNX input "0": got 8000 1  , but StimuliProvider provides 1 8000 1 "

So I'm thinking we also need to add a dimension to the input size somwhere in the Conv Cell ?

**Comment by olivierbichler-cea** [comment]
Normally using "SIze=8000 1" should be OK, even if the ONNX has one less dimension. I am confused about what is not working using this setting?
Also, I committed a patch with the modifications, as there was still an error with the code I put above...

**Comment by andreistoian** [comment]
I pulled the latest version. 
With Size=8000 1 : mInputdims of the conv cell is set to 8000 1. Then the check for 

``` if (mKernelDims.size() != mInputsDims.size() - 1) {
        std::stringstream msgStr;
        msgStr << "ConvCell::setOutputsDims(): the number of dimensions of the"
```

fails as both mKernelDims and mInputsdims are 2D. I disable the check but then the outputsize is computed wrong which triggers a failed check when unpacking the kernel tensor. I can provide the onnx and ini for debugging.

With Size=1 8000 1: i disable the check for onnx input size but I get into trouble in the output size calculation

```
OutputsDims[dim] = (unsigned int)std::ceil(
            std::floor((mInputsDims[dim]
                + 2 * mPaddingDims[dim] + mExtPaddingDims[dim]
                + mExtPaddingDims[mKernelDims.size() + dim]
                - kernelExtent + mStrideDims[dim])
                        / (double)mStrideDims[dim])
                            / (double)mSubSampleDims[dim]);
```
since mInputsdim[0] is 1 and for a kernel of size 11 this gives a negative number in the calculation triggering a floating point exception in floor.

**Comment by olivierbichler-cea** [comment]
Actually, you should try Size=8000 1 1.
I got confused because I usually use the SizeX and SizeY parameters, and I you do, the third dimension is implicitly set to 1 (NbChannels=1 by default).

**Comment by andreistoian** [comment]
Thanks, it now passes the checks for in ConvCell.cpp:994 and DeepNetGenerator.cpp:638 but fails when unpacking the weight tensors from Onnx: 

```
                        convCell->setWeight(output, channel,
                                            kernels[output][channel / group]);
```

The convCell kernel tensor has size (11,1) while the ONNX unpacked kernel has size (11), thus we get an assert in Tensor.cpp:627    
`assert(mDims.size() == tensor.nbDims());`

**Comment by olivierbichler-cea** [comment]
Ok, I think we are almost there...
Could you try replacing the line

`sharedSynapses[output][channel] = tensor_cast<T>(value);`

by 

```
        if (value.nbDims() < mKernelDims.size()) {
            for (size_t dim = 0; dim < value.nbDims(); ++dim) {
                assert(value.dims()[dim] == mKernelDims[dim]);
            }

            Tensor<T> valueND = tensor_cast<T>(value);
            valueND.reshape(std::vector<size_t>(mKernelDims.begin(),
                                                mKernelDims.end()));
            sharedSynapses[output][channel] = valueND;
        }
        else
            sharedSynapses[output][channel] = tensor_cast<T>(value);
```

in the `setWeight()` method both in ConvCell_Frame.hpp and ConvCell_Frame_CUDA.hpp?

**Comment by andreistoian** [comment]
I had to disable this test in PoolCell_Frame.cpp:70 but now the export works!!
```
    if (poolDims.size() != 2) {
        throw std::domain_error("PoolCell_Frame: only 2D pooling is"
                                " supported");
    }
```

**Comment by olivierbichler-cea** [comment]
Ok, this should now be fixed in the latest commit. Please try again to double-check, because there were some other missing changes...

**Comment by andreistoian** [comment]
It works fine now!

Thanks

**Comment by olivierbichler-cea** [comment]
Thank you, closing the issue...
